feat: fonts module

lucasvsr opened the pull request:
# `fonts` installation module

Use it to install [nerd-fonts](https://www.nerdfonts.com/font-downloads) or [google-fonts](https://fonts.google.com/). This module will run each build always downloading the latest version from properly configured fonts.

## Configuration example

```yaml

- type: fonts
  fonts:
    nerd-fonts:
      - FiraCode # don't add "Nerd Font" suffix.
      - Hack
      - SourceCodePro
      - Terminus
      - JetBrainsMono
      - NerdFontsSymbolsOnly
    google-fonts:
      - Roboto
      - Open Sans

```

lucasvsr committed:
feat: nerd-fonts modules

lucasvsr committed:
feat: nerd-fonts modules

xynydev commented:
Thank you, this is certainly very cool. I don't have time to read the code for a while now, but font installation is for sure one of the things bling was made to fix.

This makes me wonder; shouldn't we make another one for google fonts or a similar service, and if we do, should we just remove the Inter & Ubuntu font files from bling and completely rely on the module?

lucasvsr commented:
> Thank you, this is certainly very cool. I don't have time to read the code for a while now, but font installation is for sure one of the things bling was made to fix.
> 
> This makes me wonder; shouldn't we make another one for google fonts or a similar service, and if we do, should we just remove the Inter & Ubuntu font files from bling and completely rely on the module?

Thanks! I created this module to always have the latest version of the fonts, periodically looking at the nerd-fonts repository wasn't that productive...

About doing something similar with Google fonts, I can try and if I succeed, I'll open the PR here. What you think?

tulilirockz commented:
That would be really sweet! Shouldn't be too hard to adapt the code to use the Google fonts thing instead!

tulilirockz commented:
Also, yeah, we should rely just on the module instead of shipping the fonts ourselves, it'll both decrease the maintenance burden and make it more reliable for users to install those fonts

xynydev commented on modules/nerd-fonts/nerd-fonts.sh:
```diff
@@ -0,0 +1,83 @@
+
+echo "--- Downloading selected nerd-fonts ---"
+
+get_yaml_array nerdfonts '.fonts[]' "$1"
```
array name should be CAPITALIZED (convention, means that the variable is only read by code, not changed)

lucasvsr replied:
In the font module I followed this pattern

xynydev commented on modules/nerd-fonts/nerd-fonts.sh:
```diff
@@ -0,0 +1,83 @@
+
+mkcd $NERD_FONTS_DIR
+
+echo "--- Downloading selected nerd-fonts ---"
```
This info print is not necessary, build.sh  like ”running module nerd-fonts” just before.

lucasvsr replied:
I changed the text to specify which array I'm currently looping through. Is that ok?

xynydev replied:
Looks fine to me.

xynydev commented on modules/nerd-fonts/nerd-fonts.sh:
```diff
@@ -0,0 +1,83 @@
+
+        if [[ $downloaded -eq 0 ]]; then
+
+            echo "--- Downloading $file ---"
```
The ”---”’s signify the start of a block, the end being ”------” as per build.sh. This is just convention, but if you’re going to use a block inside a module, I think it should be a lever deeper with ”----” and use the same style of closing without any text inside the closing block.

This is mostly preference (of having the module output look the same), but also kind of my fault in not including more documentation around the conventions.

lucasvsr replied:
I confess that I didn't understand this comment, it's probably my weak English hahaha. See what it looks like now with the new code.

xynydev replied:
No problem, looks like you understood enough to do changes as needed.

What I meant is that output should look something like this, with the `----` block being a bit different from the `---` block used by build.sh and it would contain the output for the specific action and end with `--------` (2 times 4 dashes).
```
---- Downloading Google Fonts ----
<curl command output>
Downloaded font <fontname>
--------
```

xynydev commented on modules/nerd-fonts/nerd-fonts.sh:
```diff
@@ -0,0 +1,83 @@
+NERD_FONTS_DIR=/usr/share/fonts/nerd-fonts
+COMPRESS_TYPES=("tar.xz" "zip")
+
+mkcd() {
```
I don’t quite understand the purpose of this function, since it’s not hard to type mkdir and cd separetely and I don’t get why we should delete the directory we’re mkcd’ing to if it already exists, wiping the content which the user might want to keep,

lucasvsr replied:
I created this function to not have multiple sections with `mkdir -p <DIR> && cd <DIR>`. Just to make the code cleaner. You will see in this new version that I created other functions for this same purpose.

xynydev replied:
I get that it reduces repetition, but it also increases abstraction. It's fine, I wont make you remove it, I just usually rather repeat myself if the code that I am repeating is small and easily understandable. I get DRY (dont repeat yourself), but sometimes it's ok to repeat yourself.

You didn't answer why an existing directory is deleted, though.

xynydev commented on modules/nerd-fonts/nerd-fonts.sh:
```diff
@@ -0,0 +1,83 @@
+
+echo "--- Nerd-fonts installed ---"
+
+cd /
```
I don’t think this cd is necessary, build.sh should reset back to the config directory or smth like that when running each module. And if it doesn’t currently, that is an oversight by me and should be added in.

lucasvsr replied:
I preferred to place it to ensure that the execution of other modules would not fail.

xynydev replied:
That would be the job of build.sh, I will investigate further once I get the time.

xynydev commented on modules/nerd-fonts/README.md:
```diff
@@ -0,0 +1,20 @@
+# `nerd-fonts` installation module
```
There are some issues in this file regarding and wording differences between it and the other modules. If this module expands, though, might need to make bigger changes, so don’t worry about this yet.

xynydev reviewed (changes requested):
Overall pretty clean, some disagreements around conventions and style which you can feel free to contest. Read my next non-review comment, though, before making any changes.

I am not home RN (review quality might suffer from that), but once I am, probably sunday, I’ll be happy to work on this PR too.

xynydev commented:
Here’s an idea I had;
Assuming there’s going to be other font download sources than nerd fonts, I think it would make more sense to have them all inside one module. This would work similary to the bling module, which has a directory full of installer scripts for different parts of bling. These font installer scripts could just take in the fonts as shell arguments (instead of JSON) making them usable accross non-startingpoint projects (in Bluefin for installing Ubuntu & Inter for example).

The config could look something like this (normalized):
```yaml
type: fonts
install:
    - source: nerd-fonts
      fonts: 
           - Iosevka
           - Hack
           - SpaceMono
```
Or this (simplified):
```yaml
type: fonts
nerd-fonts:
    - Iosevka
    - Hack
    - SpaceMono
```

lucasvsr commented:
> Here’s an idea I had; Assuming there’s going to be other font download sources than nerd fonts, I think it would make more sense to have them all inside one module. This would work similary to the bling module, which has a directory full of installer scripts for different parts of bling. These font installer scripts could just take in the fonts as shell arguments (instead of JSON) making them usable accross non-startingpoint projects (in Bluefin for installing Ubuntu & Inter for example).
> 
> The config could look something like this (normalized):
> 
> ```yaml
> type: fonts
> install:
>     - source: nerd-fonts
>       fonts: 
>            - Iosevka
>            - Hack
>            - SpaceMono
> ```
> 
> Or this (simplified):
> 
> ```yaml
> type: fonts
> nerd-fonts:
>     - Iosevka
>     - Hack
>     - SpaceMono
> ```

OK! I had finished the Google-Fonts module, but I think you can do this adaptation. This PR should be closed, then?

xynydev commented:
> OK! I had finished the Google-Fonts module, but I think you can do this adaptation.

Great!

> This PR should be closed, then?

We can adapt this one too, you can make the changes yourself or (unless I already have them) give me write priviliges and I can do it after you merge the google fonts module into this PR. You've done great work for this and we don't wanna lose you from the commit history.

lucasvsr committed:
feat: fonts module

lucasvsr commented:
> > OK! I had finished the Google-Fonts module, but I think you can do this adaptation.
> 
> Great!
> 
> > This PR should be closed, then?
> 
> We can adapt this one too, you can make the changes yourself or (unless I already have them) give me write priviliges and I can do it after you merge the google fonts module into this PR. You've done great work for this and we don't wanna lose you from the commit history.

I made a commit with the new module.

xynydev commented:
Great, I'll redo the review when I have time.

General notes:
- No separate install scripts as a I suggested. Did you forget this or are you against it (please present reasons)?
- The YAML indentatation in the example configuration is broken (each sequential key is indented one step more, even though they do not nest)
- Most comments from my previous review still apply.

lucasvsr commented:
> Great, I'll redo the review when I have time.
> 
> General notes:
> 
> * No separate install scripts as a I suggested. Did you forget this or are you against it (please present reasons)?
> * The YAML indentatation in the example configuration is broken (each sequential key is indented one step more, even though they do not nest)
> * Most comments from my previous review still apply.

* I prefer to keep a single script and create a generic function. I didn't think the size of the script was worth the work.
* Oh sorry! I did not notice
* I responded to the previous comments. Sorry, I hadn't seen it.

lucasvsr committed:
fix: fonts module readme

xynydev commented:
>> No separate install scripts as a I suggested. Did you forget this or are you against it (please present reasons)?

> I prefer to keep a single script and create a generic function. I didn't think the size of the script was worth the work.

I undersrand but disagree. The purpose of having multiple scripts is maximum reausability, making it possible to use the font download scripts even if not using ublue-os/startingpoint. I understand that this will create more work, but I think it is worth it. An alternative to a generalized download function could be a download.sh script, or maybe we could just have the individual scripts run `source ../fonts.sh` to get the function. Of course, feel free to argue back and I'll get more people to give opinions.

lucasvsr committed:
chore: fonts module readme

lucasvsr committed:
chore: fonts module

lucasvsr commented:
> > > No separate install scripts as a I suggested. Did you forget this or are you against it (please present reasons)?
> 
> > I prefer to keep a single script and create a generic function. I didn't think the size of the script was worth the work.
> 
> I undersrand but disagree. The purpose of having multiple scripts is maximum reausability, making it possible to use the font download scripts even if not using ublue-os/startingpoint. I understand that this will create more work, but I think it is worth it. An alternative to a generalized download function could be a download.sh script, or maybe we could just have the individual scripts run `source ../fonts.sh` to get the function. Of course, feel free to argue back and I'll get more people to give opinions.

I understand that reusability is a good thing, but in this case I see no reason to do something so generic, since it is a specific module for downloading fonts for the Ublue project, which as far as I know is the only one that takes advantage of what we are doing here.

I even tried to create something very generic, but as I did so, I returned to the reflection I made in the paragraph above.

lucasvsr committed:
chore: fonts module

lucasvsr commented:
Hey guys!

I made a series of changes to make the script more organized and aiming for maximum reusability. I believe this way meets what @xynydev has been commenting on throughout this thread.

lucasvsr committed:
Update nerd-fonts.sh

lucasvsr committed:
Update google-fonts.sh

lucasvsr committed:
chore: scripts

xynydev commented:
> I understand that reusability is a good thing, but in this case I see no reason to do something so generic, since it is a specific module for downloading fonts for the Ublue project, which as far as I know is the only one that takes advantage of what we are doing here.

The point was that we could get this improvement to other ublue repos as well, like Bluefin, that don't use startingpoint's systems. And get Jorge interested like that, lol.

> I made a series of changes to make the script more organized and aiming for maximum reusability. I believe this way meets what @xynydev has been commenting on throughout this thread.

I love it, thanks!

xynydev committed:
docs: unify font module grammar with other modules

fixes some misc grammar mistakes and makes documentation more inline with that of other modules

Commit:
refactor: move download.sh to root directory
this removes the need of the font source scripts
to have an environment variable with path to the ./scripts directory
making them more portable

Commit:
fix: get correct download.sh path
(relative to currently running script)

Commit:
chore: rm loop making .sh in scripts/ executable
this is not needed, as download.sh was moved and is called with bash

xynydev reviewed (approved):
Ok, I made some miscellaneous changes myself and tested on my own image. It worked perfectly, thank you very much!
